cajita/src/Cajita_SparseDimPartitioner.hpp

```diff
@@ -0,0 +1,617 @@
+                        0, _workload_prefix_sum.extent( 2 ) ),
+                    KOKKOS_LAMBDA( const int k, int& update,
+                                   const bool final ) {
+                        const float val_i = prefix_sum( i, j, k );
+                        // const float val_i = _workload_prefix_sum( i, j, k );
+                        update += val_i;
+                        if ( final )
+                        {
+                            prefix_sum( i, j, k ) = update;
+                            // _workload_prefix_sum( i, j, k ) = update;
+                        }
+                    } );
+
+        // for ( int i = 0; i < _workload_per_tile.extent( 0 ); ++i )
+        //     for ( int j = 0; j < _workload_per_tile.extent( 1 ); ++j )
+        //         for ( int k = 0; k < _workload_per_tile.extent( 2 ); ++k )
+        //         {
+        //             printf( "i = %d, j = %d, k = %d, _workload_per_tile =
+        //             %d,"
+        //                     "_workload_prefix_sum = %d\n",
+        //                     i, j, k, _workload_per_tile( i, j, k ),
+        //                     _workload_prefix_sum( i, j, k ) );
+        //         }
+        // Kokkos::deep_copy( _workload_prefix_sum, prefix_sum );
+    }
+
+    void optimizePartition()
+    {
+        for ( int di = 0; di < 3; ++di )
+        {
+            auto rank = _ranks_per_dim[di];
+            auto rec_partition = _rectangle_partition_dev;
+            auto rec_mirror = Kokkos::create_mirror_view_and_copy(
+                Kokkos::HostSpace(), _rectangle_partition_dev );
+            auto prefix_sum = _workload_prefix_sum;
+
+            int dj = ( di + 1 ) % 3;
+            int dk = ( di + 2 ) % 3;
+            auto rank_j = _ranks_per_dim[dj];
+            auto rank_k = _ranks_per_dim[dk];
+            // compute average workload
+            Kokkos::View<int*, MemorySpace> ave_workload(
+                "ave_workload", _ranks_per_dim[dj] * _ranks_per_dim[dk] );
+            // printf( "In optimizePartition: before parallel_for\n" );
+            Kokkos::parallel_for(
+                Kokkos::RangePolicy<ExecSpace>( 0, _ranks_per_dim[dj] *
+                                                       _ranks_per_dim[dk] ),
+                KOKKOS_LAMBDA( uint32_t jnk ) {
+                    // printf(
+                    //     "in parallel_for, before calling "
+                    //     "compute_sub_workload to compute ave_workload - 0\n"
+                    //     );
+                    int j = static_cast<int>( jnk / rank_k );
+                    int k = static_cast<int>( jnk % rank_k );
+                    // printf(
+                    //     "in parallel_for, before calling "
+                    //     "compute_sub_workload to compute ave_workload - 1\n"
+                    //     );
+                    ave_workload( jnk ) =
+                        compute_sub_workload( dj, j, dk, k, rank, rec_partition,
```

Comment:
`rank`, `rec_partition`, and `prefix_sum` are already local copies, otherwise, I will CUDA runtime errors; `ave_workload` is locally defined; other arguments are just numbers.

Reply:
Sorry, I should have been more clear - `compute_sub_workload` missing a local copy is the failure I see

Reply:
Oh!! Well, even functions need local copies?

Reply:
My mistake - I read too fast and thought this was a View.  Making it `public` will enable capture on the device